### PR TITLE
llvm: dwo only emitted when object code emitted

### LINKED ISSUE #103932: `split-debuginfo = "unpacked"` fails in release build
When I add `split-debuginfo = "unpacked"` to my `profile.release` in Cargo.toml (on Linux), my build fails with:

```
error: failed to build archive: No such file or directory

error: could not compile `cfg-if` due to previous error
warning: build failed, waiting for other jobs to finish...
```

In debug I initially had the same problem but a `cargo clean` fixed it without requiring me to turn off `split-debuginfo`. However, release continues to be failing even after `cargo clean`.

### Meta
rustc --version
rustc 1.65.0 (897e37553 2022-11-02)
cargo --version
cargo 1.65.0 (4bc8f24d3 2022-10-20)

### PR DESCRIPTION
**Pull request by davidtwco**
Fixes #103932.

`CompiledModule` should not think a DWARF object was emitted when a bitcode-only compilation has happened, this can confuse archive file creation (which expects to create an archive containing non-existent dwo files).

r? @michaelwoerister

### CODE REVIEW

**Comment by michaelwoerister** [comment]
I'll take a good look at this tomorrow. Thanks for providing a fix so quickly, @davidtwco!

File: src/test/run-make-fulldeps/split-debuginfo/Makefile [hunk 1]
```diff
@@ -29,98 +29,252 @@ unpacked:
...
 	ls $(TMPDIR)/*.dwo && exit 1 || exit 0
+	ls $(TMPDIR)/*.dwp && exit 1 || exit 0
+	rm $(TMPDIR)/$(call BIN,foo)
+
+unpacked-lto: packed-lto-split packed-lto-single
+
+unpacked-lto-split:
+	$(RUSTC) baz.rs -g $(UNSTABLEOPTS) -Csplit-debuginfo=unpacked -Zsplit-dwarf-kind=split \
+		--crate-type=rlib -Clinker-plugin-lto
+	rm $(TMPDIR)/libbaz.rlib
```
Comment (michaelwoerister) [inline src/test/run-make-fulldeps/split-debuginfo/Makefile:197]: For consistency, would make sense to also test that no `.dwo` files were generated?

File: compiler/rustc_codegen_llvm/src/back/write.rs [hunk 2]
```diff
@@ -767,7 +767,8 @@ pub(crate) unsafe fn codegen(
 
     Ok(module.into_compiled_module(
         config.emit_obj != EmitObj::None,
-        cgcx.target_can_use_split_dwarf
+        matches!(config.emit_obj, EmitObj::ObjectCode(_))
```
Comment (michaelwoerister) [inline compiler/rustc_codegen_llvm/src/back/write.rs:770]: Maybe add a comment with an explanation about bitcode.

**Review by michaelwoerister — COMMENTED** [review]
Thanks for the PR, @davidtwco! What a test matrix 😉

I left a few nits below. r=me with those addressed.

File: compiler/rustc_codegen_llvm/src/back/write.rs (hunk 2 above)
Comment (davidtwco) [reply compiler/rustc_codegen_llvm/src/back/write.rs]: Fixed!

File: src/test/run-make-fulldeps/split-debuginfo/Makefile (hunk 1 above)
Comment (davidtwco) [reply src/test/run-make-fulldeps/split-debuginfo/Makefile:218]: Fixed!

File: src/test/run-make-fulldeps/split-debuginfo/Makefile [hunk 3]
```diff
@@ -29,98 +29,228 @@ unpacked:
...
+# - (main) `.dwo` never created
+# - (main) `.dwp` never created
+unpacked-crosscrate-single:
+	$(RUSTC) --crate-type lib $(UNSTABLEOPTS) -C split-debuginfo=unpacked \
+		-Zsplit-dwarf-kind=single -C debuginfo=2 -g bar.rs
+	ls $(TMPDIR)/*.rlib
+	ls $(TMPDIR)/*.o
+	ls $(TMPDIR)/*.dwo && exit 1 || exit 0
+	ls $(TMPDIR)/*.dwp && exit 1 || exit 0
+	$(RUSTC) --extern bar=$(TMPDIR)/libbar.rlib -Z unstable-options $(UNSTABLEOPTS) \
```
[the remark this answers is not in the document]
Comment (davidtwco) [reply src/test/run-make-fulldeps/split-debuginfo/Makefile]: Fixed!

File: src/test/run-make-fulldeps/split-debuginfo/Makefile [hunk 4]
```diff
@@ -29,98 +29,228 @@ unpacked:
...
+# - (main) `.dwo` present
+# - (main) `.dwp` never created
+unpacked-crosscrate-split:
+	$(RUSTC) --crate-type lib $(UNSTABLEOPTS) -C split-debuginfo=unpacked \
+		-Zsplit-dwarf-kind=split -C debuginfo=2 -g bar.rs
+	ls $(TMPDIR)/*.rlib
+	ls $(TMPDIR)/*.o && exit 1 || exit 0
+	ls $(TMPDIR)/*.dwo
+	ls $(TMPDIR)/*.dwp && exit 1 || exit 0
+	$(RUSTC) --extern bar=$(TMPDIR)/libbar.rlib -Z unstable-options $(UNSTABLEOPTS) \
```
[the remark this answers is not in the document]
Comment (davidtwco) [reply src/test/run-make-fulldeps/split-debuginfo/Makefile]: Fixed!

**Pull request merged**